src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingGraph.java

```diff
@@ -916,40 +918,84 @@ private MultiSampleEdge getHeaviestOutgoingEdge(final MultiDeBruijnVertex v) {
-        while ( index < maxIndex ) {
-            if ( path1[index] != path2[index] ) {
-                if ( ++mismatches > maxMismatches ) {
-                    return -1;
-                }
-                lastGoodIndex = index;
+    private Pair<Integer, Integer> bestPrefixMatch(final List<CigarElement> cigarElements, final byte[] path1, final byte[] path2) {
+        final int getMinMatchingBases = getMinMatchingBases();
+        int matchesSinceLastMismatch = 0;
+        int refIndex = 0, readIdx = 0;
+        int bestRefIndex = 0, bestReadIdx = 0;
+        // we want to paste the base before the last read base, thus we need to handle it in the next iteration when we find a mismatch
+        boolean wasLastIndexAMismatch = false;
+
+        for (final CigarElement ce : cigarElements) {
+            final int elementLength = ce.getLength();
+            switch (ce.getOperator()) {
+                case X:case EQ:case M:
+                    for (int j = 0; j < elementLength; j++, refIndex++, readIdx++)
+                        if (path1[refIndex] != path2[readIdx]) {
+                            bestReadIdx = readIdx;
+                            matchesSinceLastMismatch = 0;
+                            wasLastIndexAMismatch = true;
+                        } else {
+                            // We want the index BEFORE the mismatch on the ref so we can connect it
+                            if (wasLastIndexAMismatch) {
+                                bestRefIndex = refIndex;
+                                wasLastIndexAMismatch = false;
+                            }
+                            matchesSinceLastMismatch++;
+                        }
+                    break;
+                // If there is an insertion, set the bestReadIndex to the last base of said insertion
+                case I:
+                    wasLastIndexAMismatch = true;
+                    matchesSinceLastMismatch = 0;
+                    readIdx += elementLength;
+                    bestReadIdx = readIdx -1; // Compensate for the last iteration in a matching field being excessive
+                    break;
+                // If there is a deletion, set the bestReadIndex to the last base before the deletion
+                case D:
+                    bestReadIdx = readIdx - 1; // Compensate for the last iteration in a matching field being excessive
+                    wasLastIndexAMismatch = true;
+                    matchesSinceLastMismatch = 0;
+                    refIndex += elementLength;
+                    break;
+                case H:
+                case P:
+                    break;
+                default:
+                    throw new GATKException("The " + ce.getOperator() + " cigar element is not currently supported");
             }
-            index++;
         }
-        // if we got here then we hit the max index
-        return lastGoodIndex;
+        if (matchesSinceLastMismatch >= getMinMatchingBases) {
```

Comment:
Instead of tracking `matchesSinceLastMismatch` as you go, why not just find the two offsets and then backtrack from there in both sequences, looking for the required number of exact matches?